src/createWidgetBase.ts

```diff
@@ -212,15 +196,35 @@ const createWidget: WidgetFactory = createStateful
 				return this.properties.id;
 			},
 
-			applyChangedProperties: function(this: Widget<WidgetState, WidgetProperties>, previousProperties: WidgetProperties, currentProperties: WidgetProperties): void {
-				if (Object.keys(currentProperties).length) {
-					currentProperties['id'] = this.id;
-					this.setState(currentProperties);
+			setProperties(this: Widget<WidgetState, WidgetProperties>, properties: WidgetProperties) {
+				const internalState = widgetInternalStateMap.get(this);
+				const changedPropertyKeys = this.diffProperties(internalState.previousProperties, properties);
+				this.properties = this.assignProperties(internalState.previousProperties, properties, changedPropertyKeys);
+				if (changedPropertyKeys.length) {
+					this.emit({
+						type: 'properties:changed',
+						target: this,
+						properties: this.properties,
+						changedPropertyKeys
+					});
 				}
+				internalState.previousProperties = this.properties;
 			},
 
-			diffProperties(this: Widget<WidgetState, WidgetProperties>, previousProperties: any): string[] {
-				return Object.keys(this.properties);
+			diffProperties(this: Widget<WidgetState, WidgetProperties>, previousProperties: WidgetProperties, newProperties: WidgetProperties): string[] {
+				return Object.keys(newProperties);
+			},
+
+			assignProperties(this: Widget<WidgetState, WidgetProperties>, previousProperties: WidgetProperties, newProperties: WidgetProperties, changedPropertyKeys: string[]): WidgetProperties {
+				return assign({}, newProperties);
```

Comment:
Same as `diffProperties` this is overridden by default so cannot be covered by unit tests.